### PR TITLE
Support distributed aggregation

### PR DESCRIPTION
**Pull request by xander42280**

### CODE REVIEW

File: prover/src/provers/agg_prover.rs [hunk 1]
```diff
@@ -40,6 +47,7 @@ impl Prover<AggContext> for AggProver {
         let agg_pub_value_path = ctx.agg_pub_value_path.clone();
         let is_agg1 = ctx.is_agg_1;
         let is_agg2 = ctx.is_agg_2;
+        let output_dir = ctx.output_dir.clone();
         let _file = String::from("");
```
Comment (eigmax) [inline prover/src/provers/agg_prover.rs:51]: We can remove _file and _args.

File: prover/src/provers/agg_prover.rs [hunk 2]
```diff
@@ -87,6 +95,49 @@ impl Prover<AggContext> for AggProver {
...
+                format!("{}/common_circuit_data.json", path)
+            };
+            let _ = file::new(&common_data_file)
+                .write(&serde_json::to_vec(&wrapped_proof.common_data)?)?;
+            let verify_data_file = if path.ends_with('/') {
+                format!("{}verifier_only_circuit_data.json", path)
+            } else {
+                format!("{}/verifier_only_circuit_data.json", path)
+            };
+            let _ = file::new(&verify_data_file)
```
Comment (eigmax) [inline prover/src/provers/agg_prover.rs:131]: we don't have to receive the result, just remove `let _ = `

File: service/examples/stage.rs [hunk 3]
```diff
@@ -75,7 +75,7 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
                 .await?
                 .into_inner();
             if get_status_response.status != crate::stage_service::Status::Computing as u32 {
-                println!("get_status_response response: {:?}", response);
```
Comment (eigmax) [inline service/examples/stage.rs:78]: Plz change all the println! to log.

File: stage/src/stage.rs [hunk 4]
```diff
@@ -14,14 +15,25 @@ pub fn get_timestamp() -> u64 {
     duration_since_epoch.as_secs()
 }
 
+pub const STEP_INIT: i32 = 0;
```
Comment (eigmax) [inline stage/src/stage.rs:18]: Can we use Enum to wrap all the steps?

File: service/src/stage_worker.rs [hunk 5]
```diff
@@ -139,23 +177,13 @@ async fn run_stage_task(
...
-                            }
-                            if stage.agg_all_task.state != TASK_STATE_SUCCESS {
-                                crate::stage_service::stage_service::Status::AggError
-                            } else if stage.final_task.state != TASK_STATE_SUCCESS {
-                                return crate::stage_service::stage_service::Status::FinalError;
-                            } else {
-                                return crate::stage_service::stage_service::Status::InternalError;
-                            }
-                        }
+                    let get_status = || match stage.step {
```
Comment (eigmax) [inline service/src/stage_worker.rs:180]: wrap the Steps, and implement the `get_status` as its method.

File: stage/src/tasks/agg_task.rs [hunk 6]
```diff
@@ -37,3 +39,157 @@ impl Clone for AggAllTask {
...
+    pub node_info: String,
+
+    // depend
+    pub left: Option<String>,
+    pub right: Option<String>,
+}
+
+impl Clone for AggTask {
+    fn clone(&self) -> Self {
+        AggTask {
```
Comment (eigmax) [inline stage/src/tasks/agg_task.rs:74]: Can we create another structure, like RootProof, which consists of proof_path, pub_value_path,  is_agg. Then,  we can define 2 fields in AggTask,  left: Option<RootProof>, right: Option<RootProof>.

File: stage/src/tasks/agg_task.rs [hunk 7]
```diff
@@ -37,3 +39,157 @@ impl Clone for AggAllTask {
...
+            seg_size: prove_task.seg_size,
+            proof_id: prove_task.proof_id.clone(),
+            from_prove: true,
+            ..Default::default()
+        };
+        agg_task.set_out_path(prove_dir);
+        agg_task
+    }
+
+    pub fn init_from_two_prove_task(
```
Comment (eigmax) [inline stage/src/tasks/agg_task.rs:142]: ProveTask is RootProveTask?

File: prover/src/provers/agg_prover.rs (hunk 1 above)
Comment (xander42280) [reply prover/src/provers/agg_prover.rs:51]: okey

File: stage/src/tasks/agg_task.rs (hunk 7 above)
Comment (xander42280) [reply stage/src/tasks/agg_task.rs:142]: yes

File: prover/src/provers/agg_prover.rs (hunk 2 above)
Comment (xander42280) [reply prover/src/provers/agg_prover.rs:131]: If removed, an error will be reported
'error: written amount is not handled'

File: service/examples/stage.rs (hunk 3 above)
Comment (xander42280) [reply service/examples/stage.rs:78]: done

File: service/src/stage_worker.rs (hunk 5 above)
Comment (xander42280) [reply service/src/stage_worker.rs:180]: The error here is defined in the protocol stage and does not want to be coupled with the protocol here

File: stage/src/tasks/agg_task.rs (hunk 6 above)
Comment (xander42280) [reply stage/src/tasks/agg_task.rs:74]: done

**Review by eigmax — APPROVED** [review]

**Pull request merged**